src/lib.rs

```diff
@@ -99,5 +122,130 @@ pub fn cfg_vis(
         #default_item
     };
 
-    proc_macro::TokenStream::from(tokens)
+    Ok(proc_macro::TokenStream::from(tokens))
+}
+
+///
+/// # cfg visibility on fields
+///
+/// ## Rules
+///
+/// ## Example
+///
+/// ```rust
+/// use cfg_vis::cfg_vis_fields;
+///
+/// #[cfg_vis_fields]
+/// struct Foo {
+///     // while the target is linux, the visibility is `pub`.
+///     #[cfg_vis(target_os = "linux", pub)]
+///     foo: i32,
+/// }
+/// ```
+///
+#[proc_macro_attribute]
+pub fn cfg_vis_fields(
+    attr: proc_macro::TokenStream,
+    item: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+    if !attr.is_empty() {
+        let err = syn::Error::new(
+            Span::call_site(),
+            format!("unsupported arg \"{}\" for `cfg_vis_fields`", attr),
+        );
+        return proc_macro::TokenStream::from(err.into_compile_error());
+    }
+
+    let item = parse_macro_input!(item as syn::Item);
+
+    let toks = cfg_vis_fields_impl(item)
+        .map(|item| quote! { #item })
+        .unwrap_or_else(|err| err.to_compile_error());
+
+    proc_macro::TokenStream::from(toks)
+}
+
+fn cfg_vis_fields_impl(mut item: syn::Item) -> syn::Result<syn::Item> {
+    let fields = match &mut item {
+        syn::Item::Struct(s) => {
+            let fields = match &mut s.fields {
+                syn::Fields::Named(f) => &mut f.named,
+                syn::Fields::Unnamed(f) => &mut f.unnamed,
+                syn::Fields::Unit => {
+                    return Ok(item);
+                }
+            };
+
+            fields
+        }
+
+        syn::Item::Union(u) => &mut u.fields.named,
+        _ => {
+            return Err(syn::Error::new(
+                item.span(),
+                "`cfg_vis_fields` can only apply on struct or union",
+            ))
+        }
+    };
+
+    *fields = find_replace_cfg_vis(std::mem::take(fields))?;
+
+    Ok(item)
+}
+
+fn find_replace_cfg_vis(
+    fields: Punctuated<syn::Field, syn::Token![,]>,
+) -> syn::Result<Punctuated<syn::Field, syn::Token![,]>> {
+    let mut fields_replaced = Punctuated::new();
+    for mut field in fields {
```

Comment:
Why not use `fields.into_iter().flat_map(...).collect()` (

Reply:
because I want to early return from the loop.

Reply:
How about this within `flat_map`:
```rust
let pos = if let ... { ... } else {
    return vec![field];
}
```

Reply:
How do you deal with `guard_cfg_vis_unique(&field.attrs, false)?`

Reply:
sry, I overlooked `? `. how about the following:
```rust
fn find_replace_cfg_vis(
    fields: Punctuated<syn::Field, syn::Token![,]>,
) -> syn::Result<Punctuated<syn::Field, syn::Token![,]>> {
    fields
        .into_iter()
        .try_fold(Punctuated::new(), |mut fields, mut field| {
            if let Some(pos) = guard_cfg_vis_unique(&field.attrs, false)? {
                let attr = &field.attrs[pos].tokens;
                let CfgVisAttrArgsWithParens(CfgVisAttrArgs { cfg, vis }) = parse_quote!(#attr);

                let mut field_replaced = field.clone();
                field_replaced.attrs[pos] = parse_quote! { #[cfg(#cfg)] };
                field_replaced.vis = vis;
                fields.push(field_replaced);

                field.attrs[pos] = parse_quote! { #[cfg(not(#cfg))] };
            }
            fields.push(field);
            Ok(fields)
        })
}
```

Reply:
I prefer `for in` with `?` to `try_fold` with closure.

Reply:
Alright, but I think the two lines `fields.push(field);` can be written outside the if-let block :)

Reply:
great